ste/pacer-tokenBucketPacer.go

```diff
@@ -66,23 +85,44 @@ func newTokenBucketPacer(ctx context.Context, bytesPerSecond int64, expectedByte
 	return p
 }
 
-// RequestRightToSend function is called by goroutines to request right to send a certain amount of bytes.
+// RequestTrafficAllocation function is called by goroutines to request right to send a certain amount of bytes.
 // It controls their rate by blocking until they are allowed to proceed
-func (p *tokenBucketPacer) RequestRightToSend(ctx context.Context, bytesToSend int64) error {
-	for atomic.AddInt64(&p.atomicTokenBucket, -bytesToSend) < 0 {
+func (p *tokenBucketPacer) RequestTrafficAllocation(ctx context.Context, byteCount int64) error {
+
+	// block until tokens are available
+	for atomic.AddInt64(&p.atomicTokenBucket, -byteCount) < 0 {
+
 		// by taking our desired count we've moved below zero, which means our allocation is not available
-		// right now, so put back what we asked for, and wait
-		atomic.AddInt64(&p.atomicTokenBucket, bytesToSend)
+		// right now, so put back what we asked for, and then wait
+		atomic.AddInt64(&p.atomicTokenBucket, byteCount)
+
+		// vary the wait amount, to reduce risk of any kind of pulsing or synchronization effect, without the perf and
+		// and threadsafety issues of actual random numbers
+		totalWaitsSoFar := atomic.AddInt64(&p.atomicWaitCount, 1)
+		modifiedSleepDuration := time.Duration(float32(bucketDrainSleepDuration) * (float32(totalWaitsSoFar%10) + 5) / 10) // 50 to 150% of bucketDrainSleepDuration
```

Comment:
Another case where I could be dead wrong, but if I'm making enough transfers at once, is it possible that I could stack the layers too high and get more throughput than the limit?

If so, this method of limiting throughput may actually cause issues with larger transfers.

My suggestion for this is to handle transfers as a queue-- if something goes over the maximum amount of tokens in the bucket, wait until the bucket is full. If less, wait until the bucket has _just_ enough to push it through.

Reply:
I don't currently believe it's possible to get more throughput than the limit.  The logic being that tokens are only put into the bucket at a certain rate.  The net rate(*) at which we remove them is always <= the rate at at which they are added, so we can't go faster than the add rate.

I.e. if we remove too many, on line 93, then we immediately put back everything we removed, so our net change for that pass through the loop 0. The invariant maintained by all callers of RequestTrafficAllocation is that, for their final "take" from the bucket (the one that they do not immediately undo it) the remaining token ballance in the bucket was not negative after they removed their allocation. I.e. every single caller to RequestTrafficAllocation can sap "_I_ did not leave the bucket with a negative count". Since that applies to all callers, we know the bucket never is left with a negative count, so we know that the rate of taking tokens out is always <= the rate at which pacerBody is putting them in.

So I don't belive there's a bug here.  The token bucket implementation is a very standard solution to these types of problem, so I'm reluctant to change it to a queue.  Also, I've used variants of this code in two other apps and have never seen them go over their limits.  But, if there is indeed a bug, please talk me through it and help me see it! :-)

Reply:
Oh, yikes, for some reason my brain read "for" as "if"-- you're right, it can't bust past.

Had to give the code a good re-read until I found what I was missing.